### ISSUE TITLE
[question] How can I calculate standard deviation map of ownership?

### ISSUE THREAD

**Issue by y-ich** [lightvector/KataGo#495]
Hi.

I wonder that if you know standard deviation map of ownership, I mean standard deviation of each point in ownership, you may know where are current battle fields on board.

I saw the code but it seems not simple average of ownership of neural network output. There are many calculation of”weight”s.

Could you explain the algorithm of calculation of ownership?

Thanks.

**Comment by lightvector** [comment]
Conceptually, the algorithm is just to iterate through all nodes in the search tree with at least minVisits. Each node computes its average ownership to be the average of all of its children and itself, where averaging is proportional to visits, equal to 1 for this node's own neural net evaluation and children are proportional to their visits. Except for nodes that don't achieve minVisits, in that case you omit those children from the average.

Recently, since highthreads2 branch has been merged, all nodes keep track of "weight" instead of "visits", where weight behaves basically like fractional "visits", but it is still identical.

The reason for the messiness of weight calculation is just implementation details. The implementation tries very hard to ever have to avoid normalizing all the values. When you compute an average, you have to normalize by dividing by the total. However, you can avoid normalizing until the very end if you simply allow yourself to be unnormalized, and just keep track of what you are "supposed" to be dividing by without actually dividing by it.

You do NOT need to understand any of the weight implementation to compute standard deviation. All you need to do is mimic it exactly without understanding it, except that you replace `ownerMap[pos]` with `ownerMap[pos] * ownerMap[pos]`. That way, instead of computing average ownership, you compute average squared ownership.

In other words, you will have E(ownership^2). And the original algorithm gives you E(ownership).

Then the following identity will give you the variance:
Var(X) = E(X^2) - E(X)^2 

So you don't need to understand anything to compute what you want.

**Comment by y-ich** [comment]
@lightvector san,

Thank you for your advice!
I successfully implemented standard deviation map of ownership.

(The red region is the battle field on the famous position.)
![twitter_E3z48yVVEAQoPKr](https://user-images.githubusercontent.com/922001/121833541-8a148d80-cd07-11eb-96ea-14e3941de198.jpg)

**Comment by lightvector** [comment]
Cool! Nice graphic. :)

**Comment by lightvector** [comment]
By the way, if you think you have a clean implementation and that this might be of interest to other people to have in the API, feel free to submit a PR. Only if you wish, of course. :)

**Comment by y-ich** [comment]
It is my pleasure! I submitted a PR right now.
